packages/xo-server-perf-alert/src/index.js

```diff
@@ -672,28 +678,59 @@ ${entriesWithMissingStats.map(({ listItem }) => listItem).join('\n')}`
     }
   }
 
-  async getRrd(xapiObject, secondsAgo) {
+  async getRrd(xapiObject, secondsAgo, hostCache = {}) {
     const host = xapiObject.$type === 'host' ? xapiObject : xapiObject.$resident_on
     if (host == null) {
       return null
     }
-    // we get the xapi per host, because the alarms can check VMs in various pools
     const xapi = this._xo.getXapi(host.uuid)
-    const serverTimestamp = await getServerTimestamp(xapi, host)
-    const payload = {
-      host,
-      query: {
-        cf: 'AVERAGE',
-        host: (xapiObject.$type === 'host').toString(),
-        json: 'true',
-        start: serverTimestamp - secondsAgo,
-      },
+    if (hostCache[host.uuid] === undefined) {
+      hostCache[host.uuid] = getServerTimestamp(xapi, host)
+        .then(serverTimestamp => {
+          const payload = {
+            host,
+            query: {
+              cf: 'AVERAGE',
+              host: 'true',
+              json: 'true',
+              start: serverTimestamp - secondsAgo,
+            },
+          }
+          return xapi.getResource('/rrd_updates', payload)
+        })
+        .then(res => res.body.text())
+        .then(text => JSON5.parse(text))
     }
-    if (xapiObject.$type === 'VM') {
-      payload.vm_uuid = xapiObject.uuid
+    // reuse an ecisting /in flight query
+    const json = await hostCache[host.uuid]
```

Comment:
If the promise rejects, we probably want to at least remove it from the cache so that it the call can be triggered again.